extensions/mssql/src/objectExplorerNodeProvider/objectExplorerNodeProvider.ts

```diff
@@ -182,108 +160,125 @@ export class MssqlObjectExplorerNodeProvider extends ProviderBase implements sql
-				node = <T><any>await session.root.findNodeByPath(explorerContext.nodeInfo.nodePath, true);
+				node = <T><any>await session.rootNode.findNodeByPath(explorerContext.nodeInfo.nodePath, true);
 			}
 		}
 		return node;
 	}
 
-	private async findSessionForConnection(connectionProfile: sqlops.IConnectionProfile): Promise<Session> {
+	public findSqlClusterSessionBySqlConnProfile(connectionProfile: sqlops.IConnectionProfile): SqlClusterSession {
 		for (let session of this.sessionMap.values()) {
-			if (session.connection && await session.connection.isMatch(connectionProfile)) {
+			if (session.isMatchedSqlConnection(connectionProfile)) {
 				return session;
 			}
 		}
 		return undefined;
 	}
 }
 
-export class Session {
-	private _root: RootNode;
-	constructor(private _connection: Connection, private sessionId?: string) {
-	}
+export class SqlClusterSession {
+	private _sqlClusterConnection: SqlClusterConnection;
+	private _sqlSession: sqlops.ObjectExplorerSession;
+	private _sqlConnProfile: sqlops.IConnectionProfile;
+	private _rootNode: SqlClusterRootNode;
 
-	public get uri(): string {
-		return this.sessionId || this._connection.uri;
+	constructor(arg: SqlClusterSessionArgs) {
+		this._sqlClusterConnection = arg.sqlClusterConnection;
+		this._sqlSession = arg.sqlSession;
+		this._sqlConnProfile = arg.sqlConnProfile;
+		this._rootNode = new SqlClusterRootNode(this,
+			new TreeDataContext(arg.appContext.extensionContext, arg.changeHandler),
+			arg.sqlSession.rootNode.nodePath);
 	}
 
-	public get connection(): Connection {
-		return this._connection;
-	}
+	public get sqlClusterConnection(): SqlClusterConnection { return this._sqlClusterConnection; }
+	public get sqlSession(): sqlops.ObjectExplorerSession { return this._sqlSession; }
+	public get sqlConnProfile(): sqlops.IConnectionProfile { return this._sqlConnProfile; }
+	public get sessionId(): string { return this._sqlSession.sessionId; }
+	public get rootNode(): SqlClusterRootNode { return this._rootNode; }
 
-	public set root(node: RootNode) {
-		this._root = node;
+	public isMatchedSqlConnection(sqlConnProfile: sqlops.IConnectionProfile): boolean {
+		return this._sqlConnProfile.id === sqlConnProfile.id;
 	}
+}
 
-	public get root(): RootNode {
-		return this._root;
-	}
+interface SqlClusterSessionArgs {
+	sqlClusterConnection: SqlClusterConnection;
+	sqlSession: sqlops.ObjectExplorerSession;
+	sqlConnProfile: sqlops.IConnectionProfile;
+	appContext: AppContext;
+	changeHandler: ITreeChangeHandler;
 }
 
-class RootNode extends TreeNode {
-	private children: TreeNode[];
-	constructor(private _session: Session, private context: TreeDataContext, private nodePath: string) {
+class SqlClusterRootNode extends TreeNode {
+	private _children: TreeNode[];
+	private _session: SqlClusterSession;
+	private _treeDataContext: TreeDataContext;
+	private _nodePath: string;
+
+	constructor(sqlClusterSession: SqlClusterSession, treeDataContext: TreeDataContext, nodePath: string) {
```

Comment:
Again, please declare the `private _session: SqlClusterSession` in the constructor instead of duplicating the code. It's the standard practice in this codebase.

```suggestion
	constructor(private _session: SqlClusterSession, private _treeDataContext: TreeDataContext, private _nodePath: string) {
```

Reply:
Oh and `nodePath` -> `nodePathValue`

Reply:
fixed